### ISSUE TITLE
Props and "view" as priority lookup sources for "wired_components"

### ISSUE THREAD

**Issue by pauleveritt** [mmerickel/wired#31]
tl;dr I need a way to get more than the container in as DI sources

# Summary

The dataclass DI system sniffs fields and looks in the container for things that match. I'm working on a mythical "wired_components" where values would first come from passed-in arguments, second from a "view" in the per-request container, then lastly, the status quo.

Wedge it into the existing Injector.__call__, or put the responsibility elsewhere (e.g. make me provide a custom Injector)?

# Background

I'm doing a React-style "component" system for my "Reinvent Sphinx" project. In it, you can have a Jinja2 template that does `{{ Breadcrumbs() }}` and a `Breadcrumbs` component will be found, instantiated, and rendered.

What does the component look like? A dataclass which has the DI DSL thingy:

```python
@component()
@dataclass
class Breadcrumbs:
  parents: List[Resource]
```

But...it's not that simple. In practice, the template that calls `{{ Breadcrumbs() }}` is three levels deep in the template calling. Which means (as React saw) that passing all possible things down through intermediate components, who don't really need them, is painful. React is on its 3rd stab at machinery for this (hooks).

# Views

My system is doing a simplification...at the very top, there's a `View` that is registered against a resource. The view has a template. That template calls the components which call the components which call the components. 

That view is special. It's a dataclass whose fields become top-level values available in *any* component rendered underneath. It's the mediator between the system and the rendering. If you want to use a component that asks for `parents`, you need to assemble that in the view.

It's kind of cool...a view instance is put in the container on each request as a singleton and you can, from anywhere, do container.get(View) to get it.

Which means, the `wired.dataclasses.injector.Injector` would need this "view" provided and use the view's fields *first* when looking for things to inject. If it wasn't in the view, a component could still have it injected from the container.

# Props

...with a caveat. You'd still want the ability to say *this* one usage of a component is passed in *this* value. Or, there's a customization value that doesn't make sense for putting in the view, for example, whether to show a breadcrumb separator after the last Item.

So you might call a component with `{{ Breadcrumb(separator='>>>') }}` in one part of a loops and `{{ Breadcrumb(separator=None) }}` in another.

# Implications

Let's say this was now a straightforward, advertised feature of `wired.dataclasses`. The `__call__` could change to:

```python
    def __call__(self, container, props=None, view=None):
```

But you might dislike putting some niche semantics into the injector.

Alternatives:

- Just have one bag that is passed in, e.g. `first_args=None`, and make `wired_components` sort it out

- Make injectors pluggable, kind an Injector, and I subclass/override `__call__` to get what I want

# FTR

Just to be clear, I have a working system that uses wired but not yet wired.dataclasses DI. I am doing some hardwired field sniffing myself to get the view fields available in construction.

**Comment by pauleveritt** [comment]
I retract this proposal.

First, the "inject first from views" thing is dumb. "Trust The Container" is my new motto. If you have stuff per-request that you want, just make a container and register singletons.

For the props part, it's a pretty small change. I'll just make a PR for it.
